service: Support null (#79)

rettichschnidi opened the pull request:
Until now, null values received from a OData server get substituted by a
default value.
Users of this library can not tell apart a value which coincidentally
equals the default, and a missing (null-ed) value received from the
server.

This commit allows the user to disable the substitution of null values by
default values, retrieving a None object instead.

Questions:
- Is this an acceptable approach to implement this features?

<s>In case the maintainers agree with the direction, I will add tests and documentation.</s> (done)

phanak-sap commented:
I think it is worth doing if there is genuine need for this - but not like additional hardcoded parameter for client and service classes, but instead new, optional parameter for the Config, that is already consumed by them. Default should be false, correct, to retain current behavior.

https://github.com/SAP/python-pyodata/blob/21a99b50c3b1718e5b83109d6b4d944d06b7d1e5/pyodata/v2/model.py#L101

It is the generic issue of specific problem of "Date valid minimum value vs default value" discussed in #143

But additionally, there is another layer of if-else possibilities - it depends also if the service metadata defines the attribute to be Nullable or not - see #102. if the `retain_null=True` and the service somehow returns Null on property that has metadata Nullable = False - it should fail obviously as currently is, not pass None object back.

rettichschnidi commented:
> I think it is worth doing if there is genuine need for this - but not like additional hardcoded parameter for client and service classes, but instead new, optional parameter for the Config, that is already consumed by them. Default should be false, correct, to retain current behavior.

I changed the Client to take the (extend) Config object. However, the server does not consume a Config, but a lit of arguments instead. Changing this would break the API.

> But additionally, there is another layer of if-else possibilities - it depends also if the service metadata defines the attribute to be Nullable or not - see #102. if the `retain_null=True` and the service somehow returns Null on property that has metadata Nullable = False - it should fail obviously as currently is, not pass None object back.

Fixed it in the latest push.

filak-sap commented on docs/usage/advanced.rst:
```diff
@@ -105,3 +105,14 @@ If you need to work with many Entity Sets the same way or if you just need to pi
+.. code-block:: python
+
+    count = getattr(northwind.entity_sets, 'Employees').get_entities().count().execute()
+    print(count)
```
The documentation looks good except I am not sure how does the code sample relates to it?

rettichschnidi replied:
Not at all :face_with_spiral_eyes: - this slipped in by accident and the documentation part is still WIP. Would like to give a "real-world" example using Northwind, just need to find some entity which actually returns a null property.

rettichschnidi replied:
Fixed

filak-sap commented:
@rettichschnidi  How does this change solve #79? Does the reported need to configure the service instance with `retain_null=True`?

rettichschnidi commented:
> How does this change solve #79? 

Empty dates will return Null/None, not the default value.

> Does the reporter need to configure the service instance with `retain_null=True`?

Yes

phanak-sap commented on pyodata/v2/service.py:
```diff
@@ -1581,10 +1590,11 @@ def function_import_handler(fimport, response):
     """OData service"""
 
-    def __init__(self, url, schema, connection):
+    def __init__(self, url, schema, connection, retain_null=False):
```
I do not particularly like this API. generic parameters, followed by boolean flag.. it is not much future-proof, when need arise for fifth or sixth configuration parameter in service. 

I understand that this is getting quite entangled an messy, because Config class is in the model.py module and so far was not used by service.py at at. And it should in fact be in the root next to client or exceptions, definitely above pyodata/v2, simply because same config parameters should work IMHO in future v4 support as well. Perhaps such move will be in the 2.0.0 version, where backward compatibility can be broken.. 

But service.py is already at the moment depending on model.py: https://github.com/SAP/python-pyodata/blob/0599db299cf5823f810b7806c2a8dda1227dde6e/pyodata/v2/service.py#L20

So why not add the `config=None` to Service, instead of  `retain_null=False`.. and in client.py:72 just pass the same value as in client.py:68

https://github.com/SAP/python-pyodata/blob/0599db299cf5823f810b7806c2a8dda1227dde6e/pyodata/client.py#L68

filak-sap commented:
Could you please rebase to the latest master (I don't like merge commits).

rettichschnidi commented:
Rebased, improved commit messages and adjusted the documented example to #197.

phanak-sap commented on pyodata/client.py:
```diff
@@ -69,7 +69,7 @@ def __new__(cls, url, connection, odata_version=ODATA_VERSION_2, namespaces=None
             # create service instance based on model we have
             logger.info('Creating OData Service (version: %d)', odata_version)
-            service = pyodata.v2.service.Service(url, schema, connection)
+            service = pyodata.v2.service.Service(url, schema, connection, config=config)
```
yeah, much better :)

phanak-sap reviewed (approved):
Looks good to me. Please resolve the conflict in changelog.md, so we have nice one commit as so far and we are ready to merge to master (and release 1.9.0)
